tests/phpunit/tests/html-api/wpHtmlProcessorBreadcrumbs.php

```diff
@@ -573,4 +573,52 @@ public function test_retains_proper_bookmarks_after_seeking_back_to_closed_eleme
 			'Should have retained breadcrumbs from bookmarked location after seeking backwards to it.'
 		);
 	}
+
+	/**
+	 * Ensures that breadcrumbs are properly reported on virtual nodes.
+	 *
+	 * @dataProvider data_virtual_nodes_breadcrumbs
+	 *
+	 * @ticket 61348
+	 */
+	public function test_breadcrumbs_on_virtual_nodes( $html, $token_position, $expected_breadcrumbs ) {
+		$processor = WP_HTML_Processor::create_fragment( $html );
+
+		for ( $i = 0; $i < $token_position; $i++ ) {
+			$processor->next_token();
+		}
+
+		$this->assertEquals( $expected_breadcrumbs, $processor->get_breadcrumbs(), "Found incorrect breadcrumbs in {$html}." );
+	}
+
+	/**
+	 * Ensures that get_current_depth reports the correct depth on virtual nodes.
+	 *
+	 * @dataProvider data_virtual_nodes_breadcrumbs
+	 *
+	 * @ticket 61348
+	 */
+	public function test_depth_on_virtual_nodes( $html, $token_position, $expected_breadcrumbs ) {
+		$processor = WP_HTML_Processor::create_fragment( $html );
+
+		for ( $i = 0; $i < $token_position; $i++ ) {
+			$processor->next_token();
+		}
+
+		$this->assertSame( count( $expected_breadcrumbs ), $processor->get_current_depth(), "Found incorrect depth in {$html}." );
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array[]
+	 */
+	public static function data_virtual_nodes_breadcrumbs() {
+		return array(
+			'Implied P tag opener on unmatched closer'    => array( '</p>', 1, array( 'HTML', 'BODY', 'P' ) ),
+			'Implied heading tag closer on heading child' => array( '<h1><h2>', 2, array( 'HTML', 'BODY' ) ),
```

Comment:
One aspect that makes it harder for me to process it is that I'm not immediately sure what position 2 means here. I read it that next token was executed twice, so the processor stops at the virtual closing `h1` tag. We don't have a way to check if the node is virtual as it's private, but maybe we could add a check about the current tag and whether it's a closer or opener.

Reply:
I agree, it also makes it easier to get the tests correct 🙂

 I pushed a change to include tag name and open/close in the data arrays.